include/seastar/core/future.hh

```diff
@@ -358,6 +358,25 @@ static constexpr bool is_tuple_effectively_trivially_move_constructible_and_dest
 
 }
 
+// Helper to convert lvalue reference functions to rvalue.
+// This is used early in then()-alike entry points to convert lvalue references into
+// rvalues (which are used to recursively call back into the same entry point) to
+// avoid having to handle both lvalues and rvalues downstream: we convert them all
+// to rvalues. This helper handles both references to callable objects such as lambdas,
+// copying them, and references to functions, decaying them to function pointers.
+template <typename Func>
+static constexpr auto func_to_rvalue(Func&& func) {
+    static_assert(std::is_lvalue_reference_v<Func>, "call with lvalue reference");
+    if constexpr (std::is_function_v<std::remove_reference_t<Func>>) {
+        // return decayed function pointer
+        return +func;
+    } else {
+        // return a copy (used as an rvalue by the called)
+        return std::forward<Func>(func);
+    }
+}
```

Comment:
The semantics are very confusing (before and after the patch).

Proposal for the future: delete the lvalue-accepting overloads. If someone wants an lvalue, they can wrap it in std::ref(), or copy it, both are explicit. Requires an API_LEVEL.